### ISSUE TITLE
Export own dmaps with original  resolution, crash in DensifuPointCloud

### ISSUE THREAD

**Issue by thomas-19** [cdcseacave/openMVS#1014]
Hi @cdcseacave 

I am currently working with my own depth images acquired from a depth sensor. In Interface.h, it's mentioned that:

// See MVS::ExportDepthDataRaw() and MVS::ImportDepthDataRaw() for usage
// Example:
// - Image-resolution at which the depth-map was estimated
// - Depth-map-resolution, for now, only the same resolution as the image is supported

You recommended matching the depth map resolution with the image resolution. For my case, both my image and depth image resolutions are 1920x1440. However, when I run the densifypointcloud with the command:

"/DensifyPointCloud -i ~/Desktop/desk_test/scan/scene.mvs -o ~/Desktop/desk_test/scan/scene_densify.mvs -w ~/Desktop/desk_test/scan/ --geometric-iters 0 --postprocess-dmaps 0"

The densifypointcloud program crashes, reporting a Segmentation fault (core dumped).

Upon tracing the bug, I found it's caused by the discrepancy between the depth map and imageData resolutions in the function:

void DepthMapsData::FuseDepthMaps(PointCloud& pointcloud, bool bEstimateColor, bool bEstimateNormal)

It appears the imageData is scaled, while the depth map remains at the original resolution.

Then, I try to debug this crash. I noticed that the default --resolution-level option for DensifyPointCloud is set to 1. When I change the --resolution-level to 0, DensifyPointCloud runs successfully.

Thus, there seems to be a conflict between the recommendation in Interface.h and the default setting of DensifyPointCloud. Should we consider modifying the depth map size to match the imageData size in the DensifyPointCloud to resolve this issue?

**Comment by cdcseacave** [comment]
All you said is true, however an elegant solution is a bit more
complicated: to set the depth map at any resolution and the pipeline to
still work. If you have time to make a PR with such a change that would be
helpful.

**Comment by thomas-19** [comment]
Absolutely, I'd like to contribute to this project. Along with the depth map, I'm wondering if we should also rescale the normal map and confidence map?

 could you give some hints on where to check and resize images and depthmap? I prefer to add some size check and resize code in function bool Scene::DenseReconstruction(int nFusionMode, bool bCrop2ROI, float fBorderROI). Alternatively, I thought of appending the image and depthimage check and resize at the end of the function bool ComputeDepthMaps(DenseDepthMapData& data). However, I'm uncertain if these approaches would be optimal and I would appreciate your input on this.

**Comment by thomas-19** [comment]
Hi, @cdcseacave 

I have recently studied the DenseReconstructionEstimate() function, particularly the section concerning depthmapComputed. In the analysis, I discovered that the depthmap is actually loaded in SceneDensify.cpp at line 1325:

if (depthData.IncRef(ComposeDepthFilePath(depthData.GetView().GetID(), "dmap")) == 0)
    return;

Thus, we cannot inspect the depthmap size around depthmapComputed within DenseReconstructionEstimate(). As a result, I am contemplating modifying the process to inspect and potentially adjust the dmap size right after the depthData.IncRef operation.

While I've given this considerable thought, I am still uncertain whether my understanding is completely accurate. Any advice on this would be highly appreciated.

**Comment by cdcseacave** [comment]
Lets make clear the objectives here:
 - we want to use the existing depth-maps in a way or an other no matter what the user input is in terms of resolution
 - if the user requested resolution matches the depth-map resolution, we consider them computed and skip estimation, going straight to the fusion
 - if the user requested resolution does not match the depth-map resolution, we rescale the depth-map and proceed with estimation starting from that depth-map
This decision is done per image, and the point where this needs to happen is as I said before in `DenseReconstructionEstimate()` when each image is prepared for estimation.

**Comment by 834810269** [comment]
> Hello @thomas-19 ,
> 
> I have the same idea of utilising the measured depth map from an iphone or other depth sensors. It is much appreciated if you could answer my question.
> 
> How do you convert the depth map to .dmap?
> 
> Thanks in advance!

Same question

**Comment by cdcseacave** [comment]
this was already presented in several issues already, and there is already an example in the code, see `MVS::ExportDepthDataRaw()`

you only need to copy `Interface.h` in your code and use a similar code as in the function above with your data

**Comment by cdcseacave** [comment]
When importing external depth maps at original resolution, make sure image resolution and depth-map resolution match. The depth-map size must equal the image size recorded in the \.mvs\ scene at the working resolution level.
